### PR TITLE
[igemm_dynamic] v4r1 bwd dynamic kernel

### PR DESCRIPTION
**Pull request by carlushuang**
v4r1 backward_data dynamic kernel.
This is to enable dynamic kernel in bwd direction, with v4r1 backward_data stratagy(gridwise_gemm). Although tensor contraction seems a MUST in dynamic kernels(due to its reduce in index calculation), let me re-add tensor-contraction version in future PR.

### CODE REVIEW

File: src/conv/invokers/impl_gemm_dynamic.cpp [hunk 1]
```diff
@@ -20,53 +20,215 @@ float CallImplicitGemmDynamic(const miopen::Handle& handle,
...
+                                << w_tilda
+                                << ", h_tilda_left:"
+                                << h_tilda_left
+                                << ", w_tilda_left:"
+                                << w_tilda_left
+                                << ", h_tilda_slice:"
+                                << h_tilda_slice
+                                << ", w_tilda_slice:"
+                                << w_tilda_slice);
+#endif
```
Comment (shaojiewang) [inline src/conv/invokers/impl_gemm_dynamic.cpp:173]: Could this deprecated code section be removed?

Comment (carlushuang) [reply src/conv/invokers/impl_gemm_dynamic.cpp]: removed

File: src/conv/invokers/impl_gemm_dynamic.cpp [hunk 2]
```diff
@@ -107,6 +229,39 @@ InvokerFactory MakeImplGemmDynamicDataInvokerFactory(const ConvolutionContext& c
             };
         };
     }
+    else if(ctx.direction.IsBackwardData())
+    {
+        return [ctx](const std::vector<Kernel>& kernels) {
```
Comment (TejashShah) [inline src/conv/invokers/impl_gemm_dynamic.cpp:234]: Since you are not modifying ctx inside the lambda function, you could pass it as reference i.e return [&ctx]

File: src/conv/invokers/impl_gemm_dynamic.cpp [hunk 3]
```diff
@@ -107,6 +229,39 @@ InvokerFactory MakeImplGemmDynamicDataInvokerFactory(const ConvolutionContext& c
...
+            return [=](const Handle& handle, const boost::any& primitive_parameters) {
```
Comment (TejashShah) [inline src/conv/invokers/impl_gemm_dynamic.cpp:235]: Here too, pass variables as reference return[&]

File: src/conv/invokers/impl_gemm_dynamic.cpp [hunk 4]
```diff
@@ -107,6 +229,39 @@ InvokerFactory MakeImplGemmDynamicDataInvokerFactory(const ConvolutionContext& c
...
+                    if(handle.IsProfilingEnabled())
+                    {
+                        handle.ResetKernelTime();
+                        handle.AccumKernelTime(elapsed);
+                    }
+                }
+                else
+                {
+                    MIOPEN_THROW(
+                        "Error running dynamic implicit GEMM convolution (invalid kernel name?)");
```
Comment (TejashShah) [inline src/conv/invokers/impl_gemm_dynamic.cpp:260]: When you throw error, you could add kernel name in the string so one knows which kernel name is problematic.

File: src/conv/invokers/impl_gemm_dynamic.cpp [hunk 5]
```diff
@@ -20,53 +20,175 @@ float CallImplicitGemmDynamic(const miopen::Handle& handle,
...
-    opArgs.emplace_back(stride_w);
-    opArgs.emplace_back(dilation_h);
-    opArgs.emplace_back(dilation_w);
-    opArgs.emplace_back(pad_h);
-    opArgs.emplace_back(pad_w);
-    if(kernel_is_1x1)
+    if(ctx.direction.IsForward())
     {
-        opArgs.emplace_back(__pack0);
+        bool kernel_is_1x1 = (kernel.GetName().find("igemm_v4r1_1x1_dynamic") == 0);
```
Comment (TejashShah) [inline src/conv/invokers/impl_gemm_dynamic.cpp:25]: Please use ConvolutionContextInterpreter https://github.com/ROCmSoftwarePlatform/MIOpen/blob/16d87c1952e41b54b36fe6cb59a2a40cc3275934/src/solver/implicitgemm_util.hpp#L100
that gives hi/wi/c depending upon the forward/backward direction. You could simply invoke that code before if else block.

File: src/conv/invokers/impl_gemm_dynamic.cpp [hunk 6]
```diff
@@ -107,6 +229,39 @@ InvokerFactory MakeImplGemmDynamicDataInvokerFactory(const ConvolutionContext& c
...
+                        handle.ResetKernelTime();
+                        handle.AccumKernelTime(elapsed);
+                    }
+                }
+                else
+                {
+                    MIOPEN_THROW(
+                        "Error running dynamic implicit GEMM convolution (invalid kernel name?)");
+                }
+            };
```
Comment (TejashShah) [inline src/conv/invokers/impl_gemm_dynamic.cpp:262]: Please remove this no-op semicolon after } 
One more down below,

File: src/include/miopen/conv/invokers/impl_gemm_dynamic.hpp [hunk 7]
```diff
@@ -35,6 +35,41 @@
 namespace miopen {
 namespace conv {
 
+namespace igemm_dynamic {
+template <typename T>
+T gcd(T x, T y)
```
Comment (TejashShah) [inline src/include/miopen/conv/invokers/impl_gemm_dynamic.hpp:40]: C++17 supports gcd function. So, we need until we upgrade the compiler version. Could you please move this function to some header file include/miopen so that it could be used by other files not just impl_gemm_dynamic.

File: src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    if(!(StartsWith(device_name, "gfx900") || StartsWith(device_name, "gfx906")))
+        return false;
+
+    if(!ctx.direction.IsBackwardData())
+        return false;
+
+    if(!ctx.Is2d())
+        return false;
+
+    if(!ctx.IsFp32())
```
Comment (TejashShah) [inline src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp:146]: What's the plan to support fp16?

File: src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+    kernel.l_wk.clear();
+    kernel.l_wk.push_back(block_size);
+    kernel.l_wk.push_back(1);
+    kernel.l_wk.push_back(1);
+
+    GenerateClangDefsym(options, "ROCM_METADATA_VERSION", ctx.rmv.UseV3() ? 5 : 4);
+
+    kernel.comp_options = options.str();
+
+    MIOPEN_LOG_I2(kernel.kernel_file + ":" + kernel.kernel_name);
```
Comment (TejashShah) [inline src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp:194]: Remove this log as when you increase log level to 6/7, you will get to see this info through other logs. You won't need it

File: src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+    int pad_w       = ctx.pad_w;
+    int y           = ctx.kernel_size_h;
+    int x           = ctx.kernel_size_w;
+
+    int gcd_stride_dilation_h = conv::igemm_dynamic::gcd(stride_h, dilation_h);
+    int gcd_stride_dilation_w = conv::igemm_dynamic::gcd(stride_w, dilation_w);
+    int y_tilda     = stride_h / gcd_stride_dilation_h;
+    int x_tilda     = stride_w / gcd_stride_dilation_w;
+
+    //int y_dot = (y +  y_tilda - 1) / y_tilda;
```
Comment (TejashShah) [inline src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp:65]: Remove this commented code

File: src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+                grid_size   = (gemm_m >> 6) * (gemm_n >> 6);
+                block_size  = 64;
+                kernel_name = "igemm_bwd_gtc_bt64x64x8_tt8x8_gm2x4x2_gn2x4x2_ta1x2x1x1x4_"
+                              "4x1x1x16x1_tb1x2x1x1x4x1x1_4x1x1x16x1x1x1";
+                return true;
+            }
+        }
+        else
+        {
+            // TODO: add here!
```
Comment (TejashShah) [inline src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp:128]: Remove this else block as it is no-op until you have something to add here.

File: src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+
+static inline bool FindImplicitGemmDynamicKernelBwd(const ConvolutionContext& ctx,
+                                                    std::string& kernel_name,
+                                                    int& block_size,
+                                                    int& grid_size)
+{
+    // TODO: add more dynamic kernel to expand support range, and update this function
+    // clang-format off
+    // refer to ConvolutionContextInterpreter, in bwd most dimension is reversed
+    int hi          = ctx.out_height;
```
Comment (TejashShah) [inline src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp:44]: It is better to call ConvolutionContextInterpreter e.g. hi = ConvolutionContextInterpreter::GetInputHeightHi

https://github.com/ROCmSoftwarePlatform/MIOpen/blob/16d87c1952e41b54b36fe6cb59a2a40cc3275934/src/solver/implicitgemm_util.hpp#L131

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 3 above)
Comment (atamazov) [reply src/conv/invokers/impl_gemm_dynamic.cpp:235]: I am not sure. @DrizztDoUrden - ?

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 2 above)
Comment (DrizztDoUrden) [reply src/conv/invokers/impl_gemm_dynamic.cpp:234]: No, because it would get destroyed at the run stage or even earlier.

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 3 above)
Comment (DrizztDoUrden) [reply src/conv/invokers/impl_gemm_dynamic.cpp:235]: No, because it would get destroyed at the run stage or even earlier.

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 4 above)
Comment (DrizztDoUrden) [reply src/conv/invokers/impl_gemm_dynamic.cpp:260]: This whole file should be split to separate invokers, and no kernel name-checks should be present anywhere in the library. But for now, it is a wise suggestion, I guess.

File: src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp (hunk 11 above)
Comment (DrizztDoUrden) [reply src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp:128]: No-op would be a marginal delay in host code in find. I guess readability would be a higher priority, but I would suggest making a more clear comment here.

File: src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp (hunk 8 above)
Comment (carlushuang) [reply src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp:146]: There do have plan for gfx908+fp16, which I'm going to implement next. and according to priority currently no plan for vega20+fp16.

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 5 above)
Comment (carlushuang) [reply src/conv/invokers/impl_gemm_dynamic.cpp:25]: let me refactor code here

File: src/include/miopen/conv/invokers/impl_gemm_dynamic.hpp [hunk 13]
```diff
@@ -35,6 +35,41 @@
...
+    {
+        return gcd(x - y, y);
+    }
+    else
+    {
+        return gcd(x, y - x);
+    }
+}
+
+template <typename T>
```
[the remark this answers is not in the document]
Comment (carlushuang) [reply src/include/miopen/conv/invokers/impl_gemm_dynamic.hpp:60]: min/max is indeed not used here. Let me remove them

File: src/include/miopen/conv/invokers/impl_gemm_dynamic.hpp (hunk 7 above)
Comment (carlushuang) [reply src/include/miopen/conv/invokers/impl_gemm_dynamic.hpp:40]: I move this into `src/include/miopen/numeric.hpp`

File: src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp (hunk 9 above)
Comment (carlushuang) [reply src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp:194]: Sure

File: src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp (hunk 10 above)
Comment (carlushuang) [reply src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp:65]: Sure

File: src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp (hunk 11 above)
Comment (carlushuang) [reply src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp:128]: I'll add other kernels here to deal with support range maybe in another PR or so, so currently if no-op, this function will return false, and finally lead to no suitable solution found (which I think is desired behavior). So let me add more comment here to describe what happends.

File: src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp (hunk 12 above)
Comment (carlushuang) [reply src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp:44]: ok

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 6 above)
Comment (carlushuang) [reply src/conv/invokers/impl_gemm_dynamic.cpp:262]: Do you mean this `MIOPEN_THROW` to be put into another place outside the inner most lambda?
Not sure if that is OK, but currently the logic is the same as that in hip implicit gemm. https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/src/conv/invokers/impl_gemm.cpp#L171

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 2 above)
Comment (carlushuang) [reply src/conv/invokers/impl_gemm_dynamic.cpp:234]: seems no change needed?

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 4 above)
Comment (carlushuang) [reply src/conv/invokers/impl_gemm_dynamic.cpp:260]: I think if no kernel name check should be present, there might be a lot of re-factor work (even in HIP based igemm), so I guess current implementation is OK?

File: src/conv/invokers/impl_gemm_dynamic.cpp [hunk 14]
```diff
@@ -107,6 +229,39 @@ InvokerFactory MakeImplGemmDynamicDataInvokerFactory(const ConvolutionContext& c
...
+                if(kernel.GetName().find("igemm_bwd_gtc") == 0)
+                {
+                    std::vector<KernelInvoke> ks;
+                    std::transform(kernels.begin(),
+                                   kernels.end(),
+                                   std::back_inserter(ks),
+                                   [&](const Kernel& k) { return handle.Run(k); });
+                    float elapsed = 0;
+
+                    elapsed = CallImplicitGemmDynamic(
```
Comment (DrizztDoUrden) [inline src/conv/invokers/impl_gemm_dynamic.cpp:248]: I would suggest updating CallImplicitGemmDynamic to take invoke params in some form instead of ConvolutionContext. ConvolutionContext may contain some outdated data from find time, depending on what you would use inside. Someone may accidentally use some data from the context which may be outdated and may not be. That would lead to problems that may be hard to dissect.

Comment (carlushuang) [reply src/conv/invokers/impl_gemm_dynamic.cpp:248]: Hi @DrizztDoUrden thanks for pointing out. I've think about this, but when doing CallImplicitGemmDynamic, I only need to query from ConvolutionContext is the tensor size (`h,w,n,c,k,filter_y,filter_x,stride_y...`). So if need to replace ConvolutionContext with some kernel arg, I can think of create a data structure like below (ImplicitGemmDynamicInvokeArgCommon), and filling the member from the member of ConvolutionContext. This way, I think this structure is a little bit redundant.
Do you still think I need create this structure? Or if I'm not understand correctly.
```
struct ImplicitGemmDynamicInvokeArgCommon{
    int hi;
    int wi;
    int n;
    int k;
    int c;
    int ho;
    int wo;
    int stride_h;
    int stride_w;
    int dilation_h;
    int dilation_w;
    int pad_h;
    int pad_w;
    int y;
    int x;
};
```

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 2 above)
Comment (TejashShah) [reply src/conv/invokers/impl_gemm_dynamic.cpp:234]: @DrizztDoUrden I see now why you had to pass it by value.

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 4 above)
Comment (TejashShah) [reply src/conv/invokers/impl_gemm_dynamic.cpp:260]: @carlushuang Please add kernel name in throw message at bare minimum

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 5 above)
Comment (TejashShah) [reply src/conv/invokers/impl_gemm_dynamic.cpp:25]: @carlushuang Please raise an another PR to do that thing. As of now, this piece of code could be left as it is.

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 6 above)
Comment (TejashShah) [reply src/conv/invokers/impl_gemm_dynamic.cpp:262]: I am talking about semicolon here after curly braces ```};```

File: src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp (hunk 11 above)
Comment (TejashShah) [reply src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp]: @DrizztDoUrden I am only suggesting to remove else block. You could still make it readable by retaining // TODO comment. No reason to keep else block.

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 3 above)
Comment (TejashShah) [reply src/conv/invokers/impl_gemm_dynamic.cpp:235]: @DrizztDoUrden Are you sure about it?  If you have access to Handle, why would you loose ctx object?  are Handles owned outside ctx?

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 4 above)
Comment (carlushuang) [reply src/conv/invokers/impl_gemm_dynamic.cpp:260]: OK, let me add it

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 6 above)
Comment (carlushuang) [reply src/conv/invokers/impl_gemm_dynamic.cpp:262]: Oh you are right. Let me remove it

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 4 above)
Comment (atamazov) [reply src/conv/invokers/impl_gemm_dynamic.cpp]: Unfortunately, this is not ok.

You do not need to refactor all igemm invokers, but the new code should be written properly.

File: src/conv/invokers/impl_gemm_dynamic.cpp [hunk 15]
```diff
@@ -20,53 +20,175 @@ float CallImplicitGemmDynamic(const miopen::Handle& handle,
...
-    opArgs.emplace_back(ho);
-    opArgs.emplace_back(wo);
-    opArgs.emplace_back(stride_h);
-    opArgs.emplace_back(stride_w);
-    opArgs.emplace_back(dilation_h);
-    opArgs.emplace_back(dilation_w);
-    opArgs.emplace_back(pad_h);
-    opArgs.emplace_back(pad_w);
-    if(kernel_is_1x1)
+    if(ctx.direction.IsForward())
```
[the remark this answers is not in the document]
Comment (carlushuang) [reply src/conv/invokers/impl_gemm_dynamic.cpp:23]: OK, let me change it

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 4 above)
Comment (carlushuang) [reply src/conv/invokers/impl_gemm_dynamic.cpp]: Hi @atamazov I've seperate the invoker in different path (fwd/bwd...)

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 3 above)
Comment (DrizztDoUrden) [reply src/conv/invokers/impl_gemm_dynamic.cpp]: Yes. I mean, I have taken a significant part in designing the system of solvers and conv ctx, but you shouldn't take my word for granted anyway. I would suggest everyone making solvers at some point look where we create them and their lifecycle in general.
TL;DR: The context is created inside of find and is never stored between API calls. It contains data that may become outdated between calls. The most simple example is the workspace size and all the buffers.

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 14 above)
Comment (DrizztDoUrden) [reply src/conv/invokers/impl_gemm_dynamic.cpp]: I think you could pass there ConvDataTensors, DataInvokeParams or boost::any containing the DataInvokeParams/WrWInvokeParams (which would soon be replaced with AnyInvokeParams), depending on whether both wrw and data are supported at the same time, and fetch the data from them rather than conv context inside.
Another option is to fetch conv::ProblemDescription from ConvolutionContext, which should not contain any outdated info at any point.

File: src/conv/invokers/impl_gemm_dynamic.cpp [hunk 16]
```diff
@@ -73,45 +75,198 @@ float CallImplicitGemmDynamic(const miopen::Handle& handle,
...
+}
+
+InvokerFactory MakeImplGemmDynamicForwardInvokerFactory(const ConvolutionContext& ctx)
+{
+    return [ctx](const std::vector<Kernel>& kernels) {
+        return [=](const Handle& handle, const boost::any& primitive_parameters) {
+            const auto data_ctx = boost::any_cast<conv::DataInvokeParams>(primitive_parameters);
+            const auto& tensors = data_ctx.tensors;
+            auto kernel         = handle.Run(kernels[0]);
+            if(kernel.GetName().find("igemm_v4r1_dynamic") == 0 ||
```
Comment (DrizztDoUrden) [inline src/conv/invokers/impl_gemm_dynamic.cpp:210]: Invokers should not check on kernel names. It is much preferred to check on some data, cache some bool value, or have separate invokers for separate paths. Checking on the kernel name is a sign for that invoker should be split to multiple. This not the case for this particular one for now as I see, but in the future, you should make separate invokers for different execution paths. I am really sorry I had to make igemm shared invoker as it is, but time was the limiting factor and my puny experience with igemm solvers didn't allow me to split the invoking code at that moment, offloading that to future developments.
Avoiding the huge runtime blocks with spaghettified invocation code is part of the reason behind the invokers' system.

Comment (carlushuang) [reply src/conv/invokers/impl_gemm_dynamic.cpp:210]: @DrizztDoUrden Oh now I understand. Invokers are intent to serve one or group of kernels that have the same input argument right? If kernel arg or path is different, should separate into different invokers.
Just to make sure, in dynamic-igemm, or some other algorithms, that have multiple kernels to serve as one solver (like kernel opt for ho/wo is even, or opt for 1x1 filter, etc...). If that multiple kernel lives in the same asm or hip source file, the kernel name must be different. So in this case, the logic to choose which kernel to use should live in solver, not invoker, am I correct?

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 14 above)
Comment (carlushuang) [reply src/conv/invokers/impl_gemm_dynamic.cpp]: @DrizztDoUrden OK let me think about writing code in this way.

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 16 above)
Comment (carlushuang) [reply src/conv/invokers/impl_gemm_dynamic.cpp:210]: cc @shaojiewang please notice

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 14 above)
Comment (carlushuang) [reply src/conv/invokers/impl_gemm_dynamic.cpp]: @DrizztDoUrden Hi I choose to fetch conv::ProblemDescription from ConvolutionContext, and re-factor the invoker functions. Please help check if that is desired.

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 3 above)
Comment (carlushuang) [reply src/conv/invokers/impl_gemm_dynamic.cpp]: This capture of ctx is changed to conv::ProblemDescription

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 16 above)
Comment (carlushuang) [reply src/conv/invokers/impl_gemm_dynamic.cpp]: kernel name related check in invoker is removed

Comment (DrizztDoUrden) [reply src/conv/invokers/impl_gemm_dynamic.cpp]: Yeah, exactly.

Comment (shaojiewang) [reply src/conv/invokers/impl_gemm_dynamic.cpp]: > cc @shaojiewang please notice

Yes, got it.

**Comment by carlushuang** [comment]
> @carlushuang It looks like the code is approved for merge.
> 
> I have a few other questions related to perf
> 
> * Do you have perf comparison of resnext101/50 configs between bwd v4r1 dynamic and bwd v4r1
>   static?
> * What is current applicability criteria for bwd v4r1 dynamic vs v4r1 static? It looks like you are testing your kernel in CI with help of DYNAMIC_IMPLICITGEMM_COMMON and    MIOPEN_DEBUG_FIND_ONLY_SOLVER
>   When do we plan to either not run static kernel and how do we plan to make that choice?

Hi @TejashShah . Currently for gfx906 fp32 the dynamic kernel is within 10% performance drop compared with static kernel. currently the support range of dynamic kernel is not as good as static kernel, I'm trying to balance my work with 1) extend support range 2) performance, and 3) gfx908. Currently I'm not testing the config from resnext101/50, if you wish I can do it offline. And for the criteria to use only dynamic kernel, at current stage I think it's based on 2 factors 1) perf drop within 10%. 2) coverage, at least cover high priority model. I'm trying to make my best to make both happen for the up-comming gfx908 PR, but if not, at least make sure what I can cover can match the 1) perf drop within 10%.

**Comment by daniellowell** [comment]
@zjing14 Can you review this PR please.

**Comment by TejashShah** [comment]
> 
> 
> > @carlushuang It looks like the code is approved for merge.
> > I have a few other questions related to perf
> > 
> > * Do you have perf comparison of resnext101/50 configs between bwd v4r1 dynamic and bwd v4r1
> >   static?
> > * What is current applicability criteria for bwd v4r1 dynamic vs v4r1 static? It looks like you are testing your kernel in CI with help of DYNAMIC_IMPLICITGEMM_COMMON and    MIOPEN_DEBUG_FIND_ONLY_SOLVER
> >   When do we plan to either not run static kernel and how do we plan to make that choice?
> 
> Hi @TejashShah . Currently for gfx906 fp32 the dynamic kernel is within 10% performance drop compared with static kernel. currently the support range of dynamic kernel is not as good as static kernel, I'm trying to balance my work with 1) extend support range 2) performance, and 3) gfx908. Currently I'm not testing the config from resnext101/50, if you wish I can do it offline. And for the criteria to use only dynamic kernel, at current stage I think it's based on 2 factors 1) perf drop within 10%. 2) coverage, at least cover high priority model. I'm trying to make my best to make both happen for the up-comming gfx908 PR, but if not, at least make sure what I can cover can match the 1) perf drop within 10%.

I concur with you on the approach. For initial iteration reduction project, we focus on vega20 first, the gfx908. That's good this PR is targeted for vega20. 

If I train any resnext/inception model end-to-end with this PR, would I see the compilation reduction, by default set of environment variables or do I need to set any specific env variables.?

**Comment by carlushuang** [comment]
> If I train any resnext/inception model end-to-end with this PR, would I see the compilation reduction, by default set of environment variables or do I need to set any specific env variables.?

There is a PR:https://github.com/ROCmSoftwarePlatform/MIOpen/pull/326 that enable hybrid mode of igemm. I believe E2E run with dynamic kernel can be achieved in that mode. Because currently, the noticeable 10% perf drop of dynamic kernel may lead to it unpicked if run normally.

File: src/include/miopen/numeric.hpp [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+ *******************************************************************************/
+#ifndef GUARD_MLOPEN_NUMERIC_HPP
+#define GUARD_MLOPEN_NUMERIC_HPP
+
+#include <numeric>
+
+namespace miopen {
+
+template <typename T>
+T gcd(T x, T y)
```
Comment (atamazov) [inline src/include/miopen/numeric.hpp:34]: :anchor: possibility of stack overflow, copy-pasta
